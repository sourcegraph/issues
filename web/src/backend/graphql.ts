import { Observable } from 'rxjs'
import { GraphQLResult, requestGraphQL as requestGraphQLCommon } from '../../../shared/src/graphql/graphql'

const getHeaders = (): { [header: string]: string } => ({
    ...window.context.xhrHeaders,
    Accept: 'application/json',
    'Content-Type': 'application/json',
    'X-Sourcegraph-Should-Trace': new URLSearchParams(window.location.search).get('trace') || 'false',
})

/**
 * Does a GraphQL request to the Sourcegraph GraphQL API running under `/.api/graphql`
 *
 * @param request The GraphQL request (query or mutation)
 * @param variables A key/value object with variable values
 * @returns Observable That emits the result or errors if the HTTP request failed
 */
export const requestGraphQL = <T>(
    request: string,
    variables?: {}
): Observable<GraphQLResult<T>> =>
    requestGraphQLCommon({
        request,
        variables,
        headers: getHeaders(),
    })

/**
 * Does a GraphQL query to the Sourcegraph GraphQL API running under `/.api/graphql`
 *
 * @param request The GraphQL query
 * @param variables A key/value object with variable values
 * @returns Observable That emits the result or errors if the HTTP request failed
 */
export const queryGraphQL = <T>(request: string, variables?: {}): Observable<GraphQLResult<T>> =>
    requestGraphQLCommon({
        request,
        variables,
        headers: getHeaders(),
    })

/**
 * Does a GraphQL mutation to the Sourcegraph GraphQL API running under `/.api/graphql`
 *
 * @param request The GraphQL mutation
 * @param variables A key/value object with variable values
 * @returns Observable That emits the result or errors if the HTTP request failed
 */
export const mutateGraphQL = <T>(request: string, variables?: {}): Observable<GraphQLResult<T>> =>
    requestGraphQLCommon({
        request,
        variables,
        headers: getHeaders(),
    })
