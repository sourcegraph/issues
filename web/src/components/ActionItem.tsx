export { LinkOrButton as ActionItem } from '../../../extensions-client-common/src/ui/generic/LinkOrButton'
