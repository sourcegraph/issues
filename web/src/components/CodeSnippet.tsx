import React, { useMemo } from 'react'
import { highlightCodeSafe } from '../../../shared/src/util/markdown'

interface CodeSnippetProps {
    code: string
    language: string
}

export const CodeSnippet: React.FunctionComponent<CodeSnippetProps> = ({ code, language }) => {
    const highlightedInput = useMemo(() => ({ __html: highlightCodeSafe(code, language) }), [code, language])
    return (
        <div className="code-snippet__container rounded p-2">
            <pre className="m-0" dangerouslySetInnerHTML={highlightedInput} />
        </div>
    )
}
