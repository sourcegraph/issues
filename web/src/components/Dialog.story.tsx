import { storiesOf } from '@storybook/react'
import React, { useCallback, useState } from 'react'
import { Dialog } from './Dialog'
import { WebStory } from './WebStory'

const { add } = storiesOf('web/Dialog', module).addDecorator(story => <div className="container mt-3">{story()}</div>)

add('Interactive', () => {
    const [open, setOpen] = useState(true)
    const onClose = useCallback(() => setOpen(false), [])
    const openDialog = useCallback(() => setOpen(true), [])

    return (
        <WebStory>
            {() => (
                <div>
                    <button type="button" className="btn btn-primary" onClick={openDialog}>
                        Open modal
                    </button>
                    {open && (
                        <Dialog onClose={onClose}>
                            <div className="extension-permission-modal p-4">
                                <h1>Dialog</h1>
                                <p>You can click on the backdrop to close me</p>
                            </div>
                        </Dialog>
                    )}
                </div>
            )}
        </WebStory>
    )
})
