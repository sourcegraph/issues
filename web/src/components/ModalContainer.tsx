import CloseIcon from 'mdi-react/CloseIcon'
import React, { useCallback, useEffect, useRef } from 'react'
import classNames from 'classnames'
import { Key } from 'ts-key-enum'

interface Props {
    /** Called when user clicks outside of the modal or presses the `esc` key */
    onClose?: () => void
    hideCloseIcon?: boolean
    children: (bodyReference: React.MutableRefObject<HTMLElement | null>) => JSX.Element
    className?: string
}

export const ModalContainer: React.FunctionComponent<Props> = ({ onClose, hideCloseIcon, className, children }) => {
    const containerReference = useRef<HTMLDivElement>(null)
    const bodyReference = useRef<HTMLElement>(null)

    // On first render, close over the element that was focused to open it.
    // On unmount, refocus that element
    useEffect(() => {
        const focusedElement = document.activeElement

        containerReference.current?.focus()

        return () => {
            if (
                focusedElement &&
                (focusedElement instanceof HTMLButtonElement ||
                    focusedElement instanceof HTMLAnchorElement ||
                    focusedElement instanceof HTMLDivElement)
            ) {
                focusedElement.focus()
            }
        }
    }, [])

    // Close modal when user clicks outside of modal body
    // (optional behavior: user opts in by using `bodyReference` as the ref attribute to the body element)
    useEffect(() => {
        function handleMouseDownOutside(event: MouseEvent): void {
            const body = bodyReference.current
            if (onClose && body && body !== event.target && !body.contains(event.target as Node)) {
                onClose()
            }
        }
        // TODO(tj): Don't close modal when user mouse down inside body then mouse up outside of div?
        document.addEventListener('mousedown', handleMouseDownOutside)

        return () => document.removeEventListener('mousedown', handleMouseDownOutside)
    }, [onClose])

    // Close modal when user presses `esc` key
    const onKeyDown: React.KeyboardEventHandler<HTMLDivElement> = useCallback(
        event => {
            switch (event.key) {
                case Key.Escape: {
                    if (onClose) {
                        onClose()
                    }
                }
            }
        },
        [onClose]
    )

    return (
        <div
            ref={containerReference}
            tabIndex={-1}
            onKeyDown={onKeyDown}
            className={classNames('modal-container', className)}
        >
            <div className="modal-container__dialog">
                <div className="modal-container__close">
                    {onClose && !hideCloseIcon && (
                        <span onClick={onClose}>
                            <CloseIcon className="icon-inline btn-icon" />
                        </span>
                    )}
                </div>
                {children(bodyReference)}
            </div>
        </div>
    )
}
