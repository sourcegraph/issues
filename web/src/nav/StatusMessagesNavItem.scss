.status-messages-nav-item {
    &__dropdown-menu {
        min-width: 22rem;
        max-width: 22rem;
        padding: 0.5rem 1rem;
    }
}
