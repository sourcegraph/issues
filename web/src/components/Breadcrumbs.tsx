import React, { useState, useEffect, useMemo, useCallback } from 'react'
import ChevronRightIcon from 'mdi-react/ChevronRightIcon'
import { Link } from '../../../shared/src/components/Link'

export interface Breadcrumb {
    key: string
    element: React.ReactNode | null
    divider?: React.ReactNode
}

export interface BreadcrumbsProps {
    breadcrumbs: BreadcrumbsByDepth
}

export interface UpdateBreadcrumbsProps {
    useSetBreadcrumb: UseSetBreadcrumb
    setBreadcrumb: SetBreadcrumb
}

type BreadcrumbOrFalsy = Breadcrumb | false | null | undefined

export type UseSetBreadcrumb = (breadcrumb: BreadcrumbOrFalsy) => BreadcrumbSetters

export type SetBreadcrumb = (breadcrumb: BreadcrumbOrFalsy) => BreadcrumbSetters & { cleanup: () => void }

interface BreadcrumbSetters {
    useSetBreadcrumb: UseSetBreadcrumb
    setBreadcrumb: SetBreadcrumb
}

type BreadcrumbsByDepth = [number, BreadcrumbOrFalsy][]

/**
 *
 *
 * TODO: Document how to use `useBreadcrumbs`
 */
export const useBreadcrumbs = (): {
    breadcrumbs: BreadcrumbsByDepth
    useSetBreadcrumb: UseSetBreadcrumb
    setBreadcrumb: SetBreadcrumb
} => {
    const [breadcrumbsByDepth, setBreadcrumbsByDepth] = useState<BreadcrumbsByDepth>([
        [0, { key: 'home', element: <Link to="/search">Home</Link>, divider: null }],
    ])

    const createBreadcrumbSetters = useCallback((depth: number = 1): BreadcrumbSetters => {
        /** Shared logic between plain function and hook */
        function _internalSetBreadcrumb(breadcrumb: BreadcrumbOrFalsy): () => void {
            const entry: [number, BreadcrumbOrFalsy] = [depth, breadcrumb]

            setBreadcrumbsByDepth(breadcrumbs => [...breadcrumbs, entry])
            // cleanup
            return () => {
                setBreadcrumbsByDepth(breadcrumbs => breadcrumbs.filter(breadcrumb => breadcrumb !== entry))
            }
        }

        /** Convenience hook for function components */
        function useSetBreadcrumb(breadcrumb: BreadcrumbOrFalsy): BreadcrumbSetters {
            useEffect(() => _internalSetBreadcrumb(breadcrumb), [breadcrumb])

            return useMemo(() => createBreadcrumbSetters(depth + 1), [])
        }

        /** 'Vanilla function' for backcompat with class components */
        function setBreadcrumb(breadcrumb: BreadcrumbOrFalsy): BreadcrumbSetters & { cleanup: () => void } {
            return {
                cleanup: _internalSetBreadcrumb(breadcrumb),
                ...createBreadcrumbSetters(depth + 1),
            }
        }

        return {
            useSetBreadcrumb,
            setBreadcrumb,
        }
    }, [])

    const breadcrumbSetters = useMemo(() => createBreadcrumbSetters(), [createBreadcrumbSetters])

    return {
        breadcrumbs: breadcrumbsByDepth,
        ...breadcrumbSetters,
    }
}

/** Renders breadcrumbs by depth */
export const Breadcrumbs: React.FC<{ breadcrumbs: BreadcrumbsByDepth }> = ({ breadcrumbs }) => {
    const nodes: React.ReactNode[] = []

    breadcrumbs.sort((a, b) => a[0] - b[0])

    for (const [_depth, breadcrumb] of breadcrumbs) {
        if (breadcrumb) {
            const divider =
                breadcrumb.divider === undefined ? <ChevronRightIcon className="icon-inline" /> : breadcrumb.divider
            nodes.push(
                <span key={breadcrumb.key} className="text-muted d-flex align-items-center">
                    <span className="font-weight-semibold">{divider}</span>
                    {breadcrumb.element}
                </span>
            )
        }
    }

    return (
        <nav className="d-flex" aria-label="Breadcrumbs">
            {nodes}
        </nav>
    )
}
