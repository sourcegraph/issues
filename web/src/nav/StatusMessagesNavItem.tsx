import { startCase } from 'lodash'
import CloudAlertIcon from 'mdi-react/CloudAlertIcon'
import CloudCheckIcon from 'mdi-react/CloudCheckIcon'
import CloudSyncIcon from 'mdi-react/CloudSyncIcon'
import React from 'react'
import { ButtonDropdown, DropdownItem, DropdownMenu, DropdownToggle } from 'reactstrap'
import { Observable, SchedulerLike, Subscription, timer } from 'rxjs'
import { catchError, concatMap, map } from 'rxjs/operators'
import { Link } from '../../../shared/src/components/Link'
import { dataOrThrowErrors, gql } from '../../../shared/src/graphql/graphql'
import * as GQL from '../../../shared/src/graphql/schema'
import { asError, ErrorLike, isErrorLike } from '../../../shared/src/util/errors'
import { queryGraphQL } from '../backend/graphql'
import classNames from 'classnames'

export function fetchAllStatusMessages(): Observable<GQL.StatusMessage[]> {
    return queryGraphQL(
        gql`
            query StatusMessages {
                statusMessages {
                    __typename

                    ... on CloningStatusMessage {
                        message
                    }

                    ... on SyncErrorStatusMessage {
                        message
                        externalServiceId
                        externalServiceKind
                        externalServiceDisplayName
                    }
                }
            }
        `
    ).pipe(
        map(dataOrThrowErrors),
        map(data => data.statusMessages)
    )
}

interface StatusMessageEntryProps {
    title: string
    text: string
    showLink?: boolean
    linkTo: string
    linkText: string
    alert?: 'danger' | 'warning' | 'info' | 'success'
    linkOnClick: (event: React.MouseEvent<HTMLAnchorElement, MouseEvent>) => void
}

const StatusMessagesNavItemEntry: React.FunctionComponent<StatusMessageEntryProps> = props => (
    <div
        key={props.text}
        className={classNames('status-messages-nav-item__entry', props.alert && `alert alert-${props.alert} mb-0`)}
    >
        <h4>{props.title}</h4>
        <p>{props.text}</p>
        {props.showLink && (
            <p className="status-messages-nav-item__entry-link">
                <Link to={props.linkTo} className={props.alert ? 'alert-link' : ''} onClick={props.linkOnClick}>
                    {props.linkText}
                </Link>
            </p>
        )}
    </div>
)

interface Props {
    fetchMessages: () => Observable<GQL.StatusMessage[]>

    /** Scheduler for the refresh timer */
    scheduler?: SchedulerLike

    isSiteAdmin?: boolean
}

interface State {
    messagesOrError: GQL.StatusMessage[] | ErrorLike
    isOpen: boolean
}

const REFRESH_INTERVAL_MS = 3000

/**
 * Displays a status icon in the navbar reflecting the completion of backend
 * tasks such as repository cloning, and exposes a dropdown menu containing
 * more information on these tasks.
 */
export class StatusMessagesNavItem extends React.PureComponent<Props, State> {
    private subscriptions = new Subscription()

    public state: State = { isOpen: false, messagesOrError: [] }

    private toggleIsOpen = () => this.setState(prevState => ({ isOpen: !prevState.isOpen }))

    public componentDidMount(): void {
        this.subscriptions.add(
            timer(0, REFRESH_INTERVAL_MS, this.props.scheduler)
                .pipe(concatMap(() => this.props.fetchMessages().pipe(catchError(err => [asError(err)]))))
                .subscribe(messagesOrError => this.setState({ messagesOrError }))
        )
    }

    public componentWillUnmount(): void {
        this.subscriptions.unsubscribe()
    }

    private renderMessage(message: GQL.StatusMessage): JSX.Element | null {
        switch (message.__typename) {
            case 'CloningStatusMessage':
                return (
                    <StatusMessagesNavItemEntry
                        key={message.message}
                        title="Repositories cloning"
                        text={message.message}
                        showLink={this.props.isSiteAdmin}
                        linkTo="/site-admin/external-services"
                        linkText="Configure external services"
                        linkOnClick={this.toggleIsOpen}
                    />
                )
            case 'SyncErrorStatusMessage':
                return (
                    <StatusMessagesNavItemEntry
                        key={message.message}
                        title={`Syncing external service "${message.externalServiceDisplayName}" failed:`}
                        text={message.message}
                        showLink={this.props.isSiteAdmin}
                        linkTo={`/site-admin/external-services/${message.externalServiceId}`}
                        linkText={`Edit "${message.externalServiceDisplayName}"`}
                        linkOnClick={this.toggleIsOpen}
                        alert="danger"
                    />
                )
        }
    }

    private renderIcon(): JSX.Element | null {
        if (isErrorLike(this.state.messagesOrError)) {
            return <CloudAlertIcon className="icon-inline" />
        }
        if (this.state.messagesOrError.some(({ __typename }) => __typename === 'SyncErrorStatusMessage')) {
            return (
                <CloudAlertIcon
                    className="icon-inline"
                    data-tooltip={this.state.isOpen ? undefined : 'Syncing repositories failed!'}
                />
            )
        }
        if (this.state.messagesOrError.some(({ __typename }) => __typename === 'CloningStatusMessage')) {
            return (
                <CloudSyncIcon
                    className="icon-inline"
                    data-tooltip={this.state.isOpen ? undefined : 'Cloning repositories...'}
                />
            )
        }
        return (
            <CloudCheckIcon
                className="icon-inline"
                data-tooltip={this.state.isOpen ? undefined : 'Repositories up to date'}
            />
        )
    }

    public render(): JSX.Element | null {
        return (
            <ButtonDropdown
                isOpen={this.state.isOpen}
                toggle={this.toggleIsOpen}
                className="nav-link py-0 px-0 status-messages-nav-item__nav-link"
            >
                <DropdownToggle caret={false} className="btn btn-icon" nav={true}>
                    {this.renderIcon()}
                </DropdownToggle>

                <DropdownMenu right={true} className="status-messages-nav-item__dropdown-menu">
                    {isErrorLike(this.state.messagesOrError) ? (
                        <div className="status-messages-nav-item__entry alert alert-danger mb-0">
                            <h4>Failed to load status messages:</h4>
                            <p>{startCase(this.state.messagesOrError.message)}</p>
                        </div>
                    ) : this.state.messagesOrError.length > 0 ? (
                        this.state.messagesOrError.map((m, i) => {
                            if (!isErrorLike(this.state.messagesOrError) && i < this.state.messagesOrError.length - 1) {
                                return (
                                    <>
                                        {this.renderMessage(m)}
                                        <DropdownItem
                                            className="status-messages-nav-item__dropdown-divider"
                                            divider={true}
                                        ></DropdownItem>
                                    </>
                                )
                            }
                            return this.renderMessage(m)
                        })
                    ) : (
                        <StatusMessagesNavItemEntry
                            title="Repositories up to date"
                            text="All repositories hosted on the configured external services are cloned."
                            showLink={this.props.isSiteAdmin}
                            linkTo="/site-admin/external-services"
                            linkText="Configure external services"
                            linkOnClick={this.toggleIsOpen}
                        />
                    )}
                </DropdownMenu>
            </ButtonDropdown>
        )
    }
}
