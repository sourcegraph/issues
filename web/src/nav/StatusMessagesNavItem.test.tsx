import React from 'react'
import renderer from 'react-test-renderer'
import { from } from 'rxjs'
import { setLinkComponent } from '../../../shared/src/components/Link'
import * as GQL from '../../../shared/src/graphql/schema'
import { StatusMessagesNavItem } from './StatusMessagesNavItem'

describe('StatusMessagesNavItem', () => {
    setLinkComponent(props => <a {...props} />)
    afterAll(() => setLinkComponent(() => null)) // reset global env for other tests

    test('no messages', () => {
        const fetchMessages = () => from([])
        expect(renderer.create(<StatusMessagesNavItem fetchMessages={fetchMessages} />).toJSON()).toMatchSnapshot()
    })

    describe('one CLONING message', () => {
        const message: GQL.IStatusMessage = {
            __typename: 'StatusMessage',
            type: GQL.StatusMessageType.CLONING,
            message: 'Currently cloning repositories...',
        }

        const fetchMessages = () => {
            console.log('here')
            return from([[message]])
        }
        test('as non-site admin', () => {
            expect(renderer.create(<StatusMessagesNavItem fetchMessages={fetchMessages} />).toJSON()).toMatchSnapshot()
        })

        test('as site admin', () => {
            expect(
                renderer.create(<StatusMessagesNavItem fetchMessages={fetchMessages} isSiteAdmin={true} />).toJSON()
            ).toMatchSnapshot()
        })
    })
})
