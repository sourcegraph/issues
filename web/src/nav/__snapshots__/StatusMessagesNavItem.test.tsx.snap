// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusMessagesNavItem no messages 1`] = `
<div
  className="nav-link py-0 px-0 status-messages-nav-item__nav-link btn-group"
  onKeyDown={[Function]}
>
  <a
    aria-expanded={false}
    aria-haspopup={true}
    className="btn btn-icon nav-link"
    href="#"
    onClick={[Function]}
  >
    <svg
      className="mdi-icon icon-inline"
      data-tooltip="Repositories up to date"
      fill="currentColor"
      height={24}
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        d="M10,17L6.5,13.5L7.91,12.08L10,14.17L15.18,9L16.59,10.41M19.35,10.03C18.67,6.59 15.64,4 12,4C9.11,4 6.6,5.64 5.35,8.03C2.34,8.36 0,10.9 0,14C0,17.31 2.69,20 6,20H19C21.76,20 24,17.76 24,15C24,12.36 21.95,10.22 19.35,10.03Z"
      />
    </svg>
  </a>
  <div
    aria-hidden={true}
    className="status-messages-nav-item__dropdown-menu dropdown-menu dropdown-menu-right"
    role="menu"
    tabIndex="-1"
  >
    <div
      className="status-messages-nav-item__entry"
    >
      <h4>
        Repositories up to date
      </h4>
      <p>
        All repositories hosted on the configured external services are up to date.
      </p>
    </div>
  </div>
</div>
`;

exports[`StatusMessagesNavItem one CLONING message as non-site admin 1`] = `
<div
  className="nav-link py-0 px-0 status-messages-nav-item__nav-link btn-group"
  onKeyDown={[Function]}
>
  <a
    aria-expanded={false}
    aria-haspopup={true}
    className="btn btn-icon nav-link"
    href="#"
    onClick={[Function]}
  >
    <svg
      className="mdi-icon icon-inline"
      data-tooltip="Repositories up to date"
      fill="currentColor"
      height={24}
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        d="M10,17L6.5,13.5L7.91,12.08L10,14.17L15.18,9L16.59,10.41M19.35,10.03C18.67,6.59 15.64,4 12,4C9.11,4 6.6,5.64 5.35,8.03C2.34,8.36 0,10.9 0,14C0,17.31 2.69,20 6,20H19C21.76,20 24,17.76 24,15C24,12.36 21.95,10.22 19.35,10.03Z"
      />
    </svg>
  </a>
  <div
    aria-hidden={true}
    className="status-messages-nav-item__dropdown-menu dropdown-menu dropdown-menu-right"
    role="menu"
    tabIndex="-1"
  >
    <div
      className="status-messages-nav-item__entry"
    >
      <h4>
        Repositories up to date
      </h4>
      <p>
        All repositories hosted on the configured external services are up to date.
      </p>
    </div>
  </div>
</div>
`;

exports[`StatusMessagesNavItem one CLONING message as site admin 1`] = `
<div
  className="nav-link py-0 px-0 status-messages-nav-item__nav-link btn-group"
  onKeyDown={[Function]}
>
  <a
    aria-expanded={false}
    aria-haspopup={true}
    className="btn btn-icon nav-link"
    href="#"
    onClick={[Function]}
  >
    <svg
      className="mdi-icon icon-inline"
      data-tooltip="Repositories up to date"
      fill="currentColor"
      height={24}
      viewBox="0 0 24 24"
      width={24}
    >
      <path
        d="M10,17L6.5,13.5L7.91,12.08L10,14.17L15.18,9L16.59,10.41M19.35,10.03C18.67,6.59 15.64,4 12,4C9.11,4 6.6,5.64 5.35,8.03C2.34,8.36 0,10.9 0,14C0,17.31 2.69,20 6,20H19C21.76,20 24,17.76 24,15C24,12.36 21.95,10.22 19.35,10.03Z"
      />
    </svg>
  </a>
  <div
    aria-hidden={true}
    className="status-messages-nav-item__dropdown-menu dropdown-menu dropdown-menu-right"
    role="menu"
    tabIndex="-1"
  >
    <div
      className="status-messages-nav-item__entry"
    >
      <h4>
        Repositories up to date
      </h4>
      <p>
        All repositories hosted on the configured external services are up to date.
      </p>
      <p
        className="status-messages-nav-item__entry-link"
      >
        <a
          to="/site-admin/external-services"
        >
          Configure external services
        </a>
      </p>
    </div>
  </div>
</div>
`;
